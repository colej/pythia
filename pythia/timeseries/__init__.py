__all__=["periodograms","iterative_prewhitening"]
