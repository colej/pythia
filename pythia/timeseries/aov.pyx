import numpy as np
import array

cimport cython
cimport numpy as np

from cpython cimport array
from libc.stdio cimport printf
from libc.string cimport memset #faster than np.zeros
from libc.math cimport sqrt, fmod, log

DTYPE = np.float64
ctypedef np.float64_t DTYPE_t

ITYPE = np.intp
ctypedef np.intp_t ITYPE_t

# Wrapped helper function for squaring a value
cdef inline square(DTYPE_t x): return x*x

# Declare the structure _HistType
cdef struct _HistType:
  double HIST_SIZE;
  int sizehists;
  unsigned long *histN;
  double *histA
  double *histB
  double *histC
  double *histD
  double *histE;

# Assign _HistType structure to a ctype
ctypedef _HistType _HistType_t


"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""



# python function to call the AoV routine
def aov_periodogram(times, obs, sigma=None, min=None, max=None, frequency=True,
               oversample_factor=1., nbins=8):


  if sigma is None:
    sigma = np.ones(len(obs))*np.std(obs)

  if frequency:
    df = 1./(times.max() - times.min())
    if min is None:
      min = df * 2.

    if max is None:
      max = 0.5/(np.min(np.diff(times)))

    frequencies = np.arange(min, max, df/oversample_factor)

  else:
    dp = 0.001
    if min is None:
      min = dp

    if max is None:
      max = 25.

    frequencies = 1./np.arange(min, max, dp/oversample_factor)

  if times.ndim != 1:
    raise ValueError('times or obs should be 1-D')
  if frequencies.ndim != 1:
    raise ValueError('frequencies array should be 1-D')

  # Prepare for cython interaction
  obs         = np.asarray(obs, dtype=DTYPE, order='C')
  times       = np.asarray(times, dtype=DTYPE, order='C')
  sigmas      = np.asarray(sigma, dtype=DTYPE, order='C')
  frequencies = np.asarray(frequencies, dtype=DTYPE, order='C')


  periodogram = np.zeros(frequencies.shape, dtype=DTYPE, order='C')
  standardised_obs = np.zeros(times.shape, dtype=DTYPE, order='C')
  standardised_sigma = np.zeros(times.shape, dtype=DTYPE, order='C')

  _standardise_fused(obs, sigma, standardised_obs, standardised_sigma)
  _aov_periodogram_cython(times, standardised_obs, frequencies, periodogram, nbins)

  return frequencies.ravel(), periodogram.ravel()

"""--------------------------------------------------------------------------"""
#
# @cython.cdivision(True)
# @cython.boundscheck(False)
# @cython.wraparound(False)
# cdef _standardise_new(const DTYPE_t[::1] obs, const DTYPE_t[::1] sigma,
#                   DTYPE_t[::1] standardised_obs,
#                   DTYPE_t[::1] standardised_sigma):
#
#   cdef ITYPE_t size = obs.shape[0]
#   cdef DTYPE_t size_d = obs.shape[0]
#   cdef DTYPE_t average = 0.0
#   cdef DTYPE_t stddev = 0.0
#
#   cdef ITYPE_t i = 0
#
#   for i in range(size):
#     average += obs[i]
#     stddev += square(obs[i] - average)
#   average /= size_d
#   stddev = sqrt(stddev / size_d)
#
#   for i in range(size):
#     standardised_obs[i] = (obs[i] - average) / stddev
#     standardised_sigma[i] = sigma[i] / stddev


"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""


@cython.cdivision(True)
@cython.boundscheck(False)
@cython.wraparound(False)
cdef _standardise_fused(const DTYPE_t[::1] obs, const DTYPE_t[::1] sigma,
                        DTYPE_t[::1] standardised_obs,
                        DTYPE_t[::1] standardised_sigma):

    cdef ITYPE_t size = obs.shape[0]
    cdef DTYPE_t size_d = obs.shape[0]
    cdef DTYPE_t average = 0.0
    cdef DTYPE_t stddev = 0.0

    cdef ITYPE_t i = 0

    # Calculate the sum of observations
    for i in range(size):
        average += obs[i]

    # Calculate the average
    average /= size_d

    # Calculate the sum of squares of differences from the average
    for i in range(size):
        stddev += square(obs[i] - average)

    # Calculate the standard deviation
    stddev = sqrt(stddev / size_d)

    # Standardize the observations and sigmas
    cdef DTYPE_t temp
    for i in range(size):
        temp = (obs[i] - average) / stddev
        standardised_obs[i] = temp
        standardised_sigma[i] = sigma[i] / stddev


"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""


@cython.cdivision(True)
@cython.boundscheck(False)
@cython.wraparound(False)
cdef _aov_periodogram_cython(const DTYPE_t[::1] times, const DTYPE_t[::1] obs,
                             const DTYPE_t[::1] frequencies, DTYPE_t[::1] periodogram,
                             const ITYPE_t nbins):

  ''' Compute the AOV Periodogram using J. Devor's implementation '''

  cdef ITYPE_t size = times.shape[0]
  cdef ITYPE_t Nfreqs = frequencies.shape[0]
  cdef ITYPE_t i=0, index=0, k=0, N=0
  cdef DTYPE_t sum=0., X=0., Y=0., s1=0., L2=0., period_=0.

  cdef np.ndarray[ITYPE_t, ndim=1] histN = np.zeros(nbins, dtype=ITYPE)
  cdef np.ndarray[DTYPE_t, ndim=1] histA = np.zeros(nbins, dtype=DTYPE)
  cdef np.ndarray[DTYPE_t, ndim=1] histB = np.zeros(nbins, dtype=DTYPE)
  cdef np.ndarray[DTYPE_t, ndim=1] histC = np.zeros(nbins, dtype=DTYPE)
  cdef np.ndarray[DTYPE_t, ndim=1] histD = np.zeros(nbins, dtype=DTYPE)
  cdef np.ndarray[DTYPE_t, ndim=1] histE = np.zeros(nbins, dtype=DTYPE)

  for k in range(Nfreqs):

    sum, X, Y, s1, L2, = 0., 0., 0., 0., 0.
    period_ = 1./frequencies[k]

    memset(&histN[0], 0, sizeof(histN[0])*nbins)
    memset(&histA[0], 0, sizeof(histA[0])*nbins)
    memset(&histB[0], 0, sizeof(histB[0])*nbins)
    memset(&histC[0], 0, sizeof(histC[0])*nbins)
    memset(&histD[0], 0, sizeof(histD[0])*nbins)
    memset(&histE[0], 0, sizeof(histE[0])*nbins)

    for i in range(size):
      X = fmod(times[i],period_) / period_
      index = (<int> (nbins * X))
      Y = obs[i]
      # printf('%li\t%f\t%f\t%f\n', index, X * nbins, X, Y)

      histN[index] += 1
      histA[index] += X
      histB[index] += X*X
      histC[index] += X*Y
      histD[index] += Y
      histE[index] += Y*Y

    for i in range(nbins):
      N = histN[i]
      if N <= 1:
        periodogram[k] = 100000.0

      else:
        Y = histD[i]
        X = (Y*Y)/N

        sum += Y
        s1 += X
        L2 += (square((histC[i] * N) - (Y * histA[i])) / (N * (square(histA[i]) - (histB[i] * N)))) + histE[i] - X


    s1 -= sum * sum / size
    L2 /= size - nbins
    if ( (s1 <= 0.0) | (L2 <= 0.0) ):
      periodogram[k] = 100000.0
    else:
      periodogram[k] = s1/L2
      # periodogram[k] = log(L2/s1)


"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""
