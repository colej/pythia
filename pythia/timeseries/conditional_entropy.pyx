import numpy as np
import array

cimport cython
cimport numpy as np

from cpython cimport array
from libc.stdio cimport printf
from libc.string cimport memset #faster than np.zeros
from libc.math cimport sqrt, fmod, log

np.import_array()

DTYPE = np.float64
ctypedef np.float64_t DTYPE_t

ITYPE = np.intp
ctypedef np.intp_t ITYPE_t

cdef index_sum(const DTYPE_t[::1] arr):
    cdef DTYPE_t result = 0.0
    cdef Py_ssize_t i, n = arr.shape[0]
    for i in range(n):
        result += arr[i]
    return result

def ce_periodogram(times, obs, sigma=None, min=None, max=None, frequency=True,
                   oversample_factor=1., phase_bins=10, mag_bins=5):


  if sigma is None:
    sigma = np.ones(len(obs))*np.std(obs)

  if frequency:
    df = 1./(times.max() - times.min())
    if min is None:
      min = df * 2.

    if max is None:
      max = 0.5/(np.min(np.diff(times)))

    frequencies = np.arange(min, max, df/oversample_factor)

  else:
    dp = 0.001
    if min is None:
      min = dp

    if max is None:
      max = 25.

    frequencies = 1./np.arange(min, max, dp/oversample_factor)


  if times.ndim != 1:
    raise ValueError('times must be 1-D')
  if obs.ndim != 1:
    raise ValueError('obs must be 1-D')
  if frequencies.ndim != 1:
    raise ValueError('frequencies array should be 1-D')
  if obs.shape != times.shape:
    raise Exception("times doesn't have the same shape as obs")


  # Prepare for cython interaction
  obs         = np.asarray(obs, dtype=DTYPE, order='C')
  times       = np.asarray(times, dtype=DTYPE, order='C')
  sigmas      = np.asarray(sigma, dtype=DTYPE, order='C')
  frequencies = np.asarray(frequencies, dtype=DTYPE, order='C')

  # Declare empty periodogram to be filled
  periodogram = np.zeros(frequencies.shape, dtype=DTYPE, order='C')

  ## Call periodogram function
  _conditional_entropy_periodogram_cython(times, obs, frequencies, periodogram,
                                          phase_bins, mag_bins)

  return frequencies.ravel(), periodogram.ravel()


"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""


@cython.cdivision(True)
@cython.boundscheck(False)
@cython.wraparound(False)
cdef _conditional_entropy_periodogram_cython(const DTYPE_t[::1] times,
                                             const DTYPE_t[::1] obs,
                                             const DTYPE_t[::1] frequencies,
                                             DTYPE_t[::1] periodogram,
                                             const ITYPE_t phase_bins,
                                             const ITYPE_t mag_bins):
    cdef int j = 0
    cdef int n_freqs = frequencies.shape[0]

    for j in range(n_freqs):
        periodogram[j] = conditional_entropy_freq(times, obs, frequencies[j],
                                                  phase_bins, mag_bins)



"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""


@cython.cdivision(True)
@cython.boundscheck(False)
@cython.wraparound(False)
cdef conditional_entropy_freq(const DTYPE_t[::1] x,
                              const DTYPE_t[::1] y,
                              const DTYPE_t freq,
                              const ITYPE_t phase_bins,
                              const ITYPE_t mag_bins):

    cdef ITYPE_t ix=0, iy=0, iz=0, ii=0, ij=0, ik=0
    cdef ITYPE_t npositive_counts=0
    cdef DTYPE_t result = 0., total_counts = 0., countsPhi = 0., pPhi=0.
    cdef DTYPE_t ent = 0.
    cdef np.ndarray[DTYPE_t, ndim=2] h = np.zeros((phase_bins, mag_bins), dtype=DTYPE)
    cdef np.ndarray[DTYPE_t, ndim=1] counts
    cdef np.ndarray[DTYPE_t, ndim=1] positive_counts
    cdef np.ndarray[DTYPE_t, ndim=1] normalized_positive_counts

    cdef np.ndarray[DTYPE_t, ndim=1] edges1 = np.empty(1, dtype=DTYPE)
    cdef np.ndarray[DTYPE_t, ndim=1] edges2 = np.empty(1, dtype=DTYPE)
    cdef np.ndarray[DTYPE_t, ndim=2] edges  = np.empty((phase_bins+1, mag_bins+1), dtype=DTYPE)

    h, edges1, edges2 = _phase_diagram_histogram(x, y, freq, phase_bins, mag_bins)


    total_counts = np.sum(h)
    # index_sum(h, total_counts)

    for ix in range(phase_bins):
        # countsPhi = np.sum(h[ix,:])
        countsPhi = index_sum(h[ix,:])

        if countsPhi > 0.:
            pPhi = countsPhi/total_counts
            counts = h[ix,:]
            positive_counts = counts[counts > 0]

            if len(positive_counts) > 0:
                normalized_positive_counts = positive_counts/total_counts

                ent = np.sum(normalized_positive_counts * np.log(pPhi/normalized_positive_counts))
                result += ent


    return result

"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""

#
# @cython.cdivision(True)
# @cython.boundscheck(False)
# @cython.wraparound(False)
# cdef _phase_diagram_normalized(const DTYPE_t[::1] x,
#                                const DTYPE_t[::1] y,
#                                const DTYPE_t period,
#                                DTYPE_t[::1] xout,
#                                DTYPE_t[::1] yout):
#     cdef ITYPE_t i
#     cdef DTYPE_t x_min, y_min, x_max, y_max
#
#     x_min = x[0]%period
#     y_min = y[0]
#
#     # populate xout and yout arrays and
#     # determine min and max values
#     for i in range(x.shape[0]):
#         xout[i] = x[i] % period
#         yout[i] = y[i]
#
#     for i in range(x.shape[0]):
#         if xout[i] < x_min:
#             x_min = xout[i]
#         if yout[i] < y_min:
#             y_min = yout[i]
#
#
#     for i in range(x.shape[0]):
#         xout[i] = xout[i] - x_min
#         yout[i] = yout[i] - y_min
#
#     x_max = xout[0]
#     y_max = yout[0]
#
#     for i in range(x.shape[0]):
#         if xout[i] > x_max:
#             x_max = xout[i]
#         if y[i] > y_max:
#             y_max = yout[i]
#
#     for i in range(x.shape[0]):
#         xout[i] = xout[i] / x_max
#         yout[i] = yout[i] / y_max



"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""

@cython.cdivision(True)
@cython.boundscheck(False)
@cython.wraparound(False)
cdef _phase_diagram_normalized(const DTYPE_t[::1] x,
                               const DTYPE_t[::1] y,
                               const DTYPE_t period,
                               DTYPE_t[::1] xout,
                               DTYPE_t[::1] yout):

    # cdef ITYPE_t i
    cdef Py_ssize_t i = 0 , n = x.shape[0]
    cdef DTYPE_t x_min, y_min, x_max, y_max

    cdef np.ndarray[DTYPE_t, ndim=1] xph = np.zeros(n, dtype=DTYPE)

    # Determine min and max values
    x_min = x[0] % period
    y_min = y[0]
    x_max = x[0] % period
    y_max = y[0]

    for i in range(1,n):
        xph[i] = x[i]%period
        if xph[i] < x_min:
            x_min = xph[i]
        if xph[i] > x_max:
            x_max = xph[i]
        if y[i] < y_min:
            y_min = y[i]
        if y[i] > y_max:
            y_max = y[i]

    # Normalize arrays
    for i in range(n):
        xout[i] = (xph[i] - x_min) / (x_max - x_min)
        yout[i] = (y[i] - y_min) / (y_max - y_min)


"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""


@cython.cdivision(True)
@cython.boundscheck(False)
@cython.wraparound(False)
cdef _phase_diagram_histogram(const DTYPE_t[::1] x,
                             const DTYPE_t[::1] y,
                             const DTYPE_t freq,
                             const ITYPE_t phase_bins,
                             const ITYPE_t mag_bins):

    cdef ITYPE_t ii = 0, ij = 0
    cdef ITYPE_t nobs = x.shape[0]
    cdef np.ndarray[ITYPE_t, ndim=1] h_shape = np.empty(2, dtype=ITYPE)
    cdef np.ndarray[DTYPE_t, ndim=2] h #= np.empty((phase_bins, mag_bins), dtype=DTYPE)
    cdef np.ndarray[DTYPE_t, ndim=1] edges1 = np.empty(1, dtype=DTYPE)
    cdef np.ndarray[DTYPE_t, ndim=1] edges2 = np.empty(1, dtype=DTYPE)
    cdef np.ndarray[DTYPE_t, ndim=2] edges  = np.empty((phase_bins+1, mag_bins+1), dtype=DTYPE)
    cdef np.ndarray[ITYPE_t, ndim=2] hrange = np.zeros((2,2), dtype=ITYPE)
    cdef np.ndarray[DTYPE_t, ndim=1] phi    = np.zeros(x.shape[0], dtype=DTYPE)
    cdef np.ndarray[DTYPE_t, ndim=1] ny     = np.zeros(x.shape[0], dtype=DTYPE)
    cdef np.ndarray[DTYPE_t, ndim=2] hist_input = np.empty((nobs,2), dtype=DTYPE)

    # populate phi and ny arrays
    _phase_diagram_normalized(x, y, 1./freq, phi, ny)

    # populate hist_input array with output from the
    # _phase_diagram_normalized routine
    for ii in range(x.shape[0]):
        hist_input[ii,0] = phi[ii]
        hist_input[ii,1] = ny[ii]


    # Establish the normalisation ranges for the histogram routine
    hrange[0,1] = 1
    hrange[1,1] = 1

    # Declare h_shape for bins
    h_shape[0] = phase_bins
    h_shape[1] = mag_bins

    h, (edges1, edges2) = np.histogramdd(hist_input, bins=h_shape, range=hrange, weights=None, density=False)

    return h, edges1, edges2

"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""
"""--------------------------------------------------------------------------"""

# @cython.cdivision(True)
# @cython.boundscheck(False)
# @cython.wraparound(False)
# cdef histdd(np.ndarray[np.double_t, ndim=2] x, np.ndarray[np.intp_t, ndim=1] bins,
#             np.ndarray[np.double_t, ndim=2] range=None, np.ndarray[np.double_t, ndim=2] weights=None):
#
#     cdef np.ndarray[np.double_t, ndim=2] edges = np.zeros((len(bins), 2), dtype=np.double)
#     cdef np.ndarray[np.double_t, ndim=2] counts = np.zeros([b for b in bins], dtype=np.double)
#     cdef np.ndarray[np.intp_t, ndim=1] indices = np.empty(len(bins), dtype=np.intp)
#     cdef np.ndarray[np.intp_t, ndim=1] strides = np.empty(len(bins), dtype=np.intp)
#     cdef int i, j, n, m, k
#
#     # Set default range to the data range
#     if range is None:
#         range = np.zeros((len(bins), 2), dtype=np.double)
#         for i in range(x.shape[1]):
#             range[i, 0] = x[:, i].min()
#             range[i, 1] = x[:, i].max()
#
#     # Create bins and edges for each dimension
#     for i in range(x.shape[1]):
#         if bins[i] < 1:
#             raise ValueError('`bins` must be positive integers')
#         edges[i] = np.linspace(range[i, 0], range[i, 1], bins[i]+1)
#
#     # Compute the bin indices for each data point
#     for n in range(x.shape[0]):
#         for i in range(x.shape[1]):
#             indices[i] = np.searchsorted(edges[i], x[n, i], side='right') - 1
#         if np.all(indices >= 0) and np.all(indices < bins):
#             if weights is None:
#                 counts[tuple(indices)] += 1
#             else:
#                 counts[tuple(indices)] += weights[n]
#
#     return counts, edges

# @cython.cdivision(True)
# @cython.boundscheck(False)
# @cython.wraparound(False)
# def histogramdd(double[:, :] data, double[:, :] bins, double[:, :] range):
#     cdef int n_dims = data.shape[1]
#     cdef int n_samples = data.shape[0]
#     cdef int i, j, k
#     cdef int[:, :] indices = <int[:, :]>np.zeros((n_samples, n_dims), dtype=np.intp)
#     cdef double[:, :] histogram = <double[:, :]>np.zeros([int(bins[i, j]) for i in range(n_dims)], dtype=np.float64)
#     cdef double step_size
#
#     for i in range(n_dims):
#         step_size = (range[i, 1] - range[i, 0]) / bins[i, 0]
#         for j in range(n_samples):
#             indices[j, i] = int((data[j, i] - range[i, 0]) // step_size)
#     for j in range(n_samples):
#         histogram[tuple(indices[j, :])] += 1.0
#     return histogram
