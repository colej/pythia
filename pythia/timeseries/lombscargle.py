import numpy as np

from astropy.timeseries import LombScargle

def normalise_distribution(s1, times):
    s1_var = np.var(s1)
    return s1/s1_var

def normalise_amplitude(s1, times):
    norm_factor = np.sqrt( 1. / len(times))
    return np.sqrt(s1) * 2. * norm_factor

def normalise_power_density(s1, times):
    T = times.ptp()
    norm_factor = np.sqrt(4.0 / len(times))
    return s1 * norm_factor**2 * T

def normalise_power(s1, times):
    return s1


def LS_periodogram(times, obs, sigma=None, min=None, max=None,
                   oversample_factor=10.0, normalisation='amplitude'):

    """
    Calculates the amplitude spectrum of a given observational signal

    Parameters
    ----------
        t : `array`
            Time values
        obs : `array`
            Flux or magnitude measurements
            This is not automatically mean-substracted!!!
        f0 : float (default None)
            Minimum frequency to calculate spectrum. Defaults to df
        fn : float
            Maximum frequency to calculate spectrum. Defaults to Nyquist.
        oversample_factor : float
            Amount by which to oversample the spectrum. Defaults to 10.
        normalisation : str (default amplitude)
            Normalise the periodogram. Options are 'distribution',
            'ampliude', 'power_density', 'power'
    """

    df = 1.0 / (times.max() - times.min())

    if min is None:
        min = df
    if max is None:
        max = 0.5 / np.median(np.diff(times))  # *nyq_mult

    freq  = np.arange(min, max, df / oversample_factor)
    ls_   = LombScargle(times, obs, dy=sigma)
    sc    = ls_.power(freq, method="fast", normalization="psd")

    if sigma is None:
        sigma = np.array([1.]) #np.ones_like(signal)
    w     = (sigma ** -2.0).sum()
    power = (1. * sc) / w

    ## Normalised to return desired output units
    noramlise_str = 'normalise_{}'.format(normalisation)
    LS_out = eval(noramlise_str+'(power,times)')

    return freq, LS_out
