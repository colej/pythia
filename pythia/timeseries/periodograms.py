import numpy as np
from astropy.timeseries import LombScargle


def normalise_distribution(s1, times):
    s1_var = np.var(s1)
    return s1/s1_var

def normalise_amplitude(s1, times):
    norm_factor = np.sqrt( 1. / len(times))
    return np.sqrt(s1) * 2. * norm_factor

def normalise_power_density(s1, times):
    T = times.ptp()
    norm_factor = np.sqrt(4.0 / len(times))
    return s1 * norm_factor**2 * T

def normalise_power(s1, times):
    return s1


def LS_periodogram(times, signal, dy=None, f0=None, fn=None,
                   oversample_factor=10.0, normalisation='amplitude'):

    """
    Calculates the amplitude spectrum of a given signal

    Parameters
    ----------
        t : `array`
            Time values
        y : `array`
            Flux or magnitude measurements
            This is not automatically mean-substracted!!!
        f0 : float (default None)
            Minimum frequency to calculate spectrum. Defaults to df
        fn : float
            Maximum frequency to calculate spectrum. Defaults to Nyquist.
        oversample_factor : float
            Amount by which to oversample the spectrum. Defaults to 10.
        normalisation : str (default amplitude)
            Normalise the periodogram. Options are 'distribution',
            'ampliude', 'power_density', 'power'
    """

    df = 1.0 / (times.max() - times.min())

    if f0 is None:
        f0 = df
    if fn is None:
        fn = 0.5 / np.median(np.diff(times))  # *nyq_mult

    freq  = np.arange(f0, fn, df / oversample_factor)
    ls_   = LombScargle(times, signal, dy=dy)
    sc    = ls_.power(freq, method="fast", normalization="psd")

    if dy is None:
        dy = np.array([1.]) #np.ones_like(signal)
    w     = (dy ** -2.0).sum()
    power = (1. * sc) / w

    ## Normalised to return desired output units
    noramlise_str = 'normalise_{}'.format(normalisation)
    LS_out = eval(noramlise_str+'(power,times)')

    return freq, LS_out
