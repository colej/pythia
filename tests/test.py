import numpy as np
import matplotlib.pyplot as plt

from matplotlib import rc

from pythia.timeseries.lombscargle import LS_periodogram
from pythia.timeseries.iterative_prewhitening import run_ipw


plt.rcParams.update({
    "text.usetex": True,
        "font.family": "sans-serif",
    "font.sans-serif": ["Helvetica"]})
plt.rcParams.update({
    "pgf.rcfonts": False,
    "pgf.texsystem": "pdflatex",
    "pgf.preamble": "\n".join([
         r"\usepackage{amsmath}",
         r"\usepackage[utf8x]{inputenc}",
         r"\usepackage[T1]{fontenc}",
         r"\usepackage{cmbright}",
    ]),
})

plt.rcParams['xtick.labelsize']=18
plt.rcParams['ytick.labelsize']=18
if __name__=="__main__":

  # times,fluxes = np.loadtxt("./test.data",unpack=True)
  times,fluxes = np.loadtxt("./test_tess.dat",unpack=True)
  nu_w,amp_w = LS_periodogram(times,np.ones_like(times),min=0.001,max=5.)
  nu,amp = LS_periodogram(times,fluxes-np.mean(fluxes),max=50.)


  figw,axw = plt.subplots(1,1,figsize=(6.6957,6.6957),num=2)
  axw.plot(nu_w,amp_w,'k-')
  axw.plot(-1.*nu_w,amp_w,'k-')
  axw.set_xlabel(r'$\nu\,\,{\rm [d^{-1}]}$',fontsize=14)
  axw.set_ylabel(r'${\rm Amplitude\,\,[ppm]}$',fontsize=14)
  axw.set_title(r'${\rm Window}$',fontsize=14)



  fig,ax = plt.subplots(1,1,figsize=(6.6957,6.6957),num=1)
  ax.plot(nu,amp,'k-')
  ax.set_xlabel(r'$\nu\,\,{\rm [d^{-1}]}$',fontsize=14)
  ax.set_ylabel(r'${\rm Amplitude\,\,[ppm]}$',fontsize=14)
  ax.set_title(r'${\rm Example}$',fontsize=14)
  plt.show()

  yerr = 0.0005* np.ones_like(times)
  residuals, model, offsets, \
  frequencies, amplitudes, phases,\
  stop_criteria, noise_level = run_ipw(times,fluxes-np.mean(fluxes), yerr, maxiter=5, max=30.)

  np.savetxt('test.out',np.array([offsets,frequencies,amplitudes,phases]).T)
  fig,ax = plt.subplots(1,1,figsize=(6.6957,6.6957))

  print(' C + A*sin( 2*pi*f*(t-t0)+phi )')
  nu_,amp_ = LS_periodogram(times, residuals, max=6.5, normalisation='amplitude')

  print(offsets)
  print(frequencies)
  print(amplitudes)
  print(phases)
  print(stop_criteria)

  outstr = '{} -- C: {:.6f} -- A: {:.6f} -- f: {:.6f} -- phi: {:.6f} -- SNR: {:.6f}'
  for ii,freq in enumerate(frequencies):
      print(outstr.format(ii, offsets[ii], amplitudes[ii], freq, phases[ii],stop_criteria[ii]))
      ax.axvline(freq, linestyle='--',color='red',alpha=0.4)
      ax.axhline(amplitudes[ii], linestyle='--',color='red',alpha=0.4)

  ax.plot(nu,amp,'-',color='grey')
  ax.plot(nu_,amp_,'k-')
  ax.set_xlabel(r'$\nu\,\,{\rm [d^{-1}]}$',fontsize=14)
  ax.set_ylabel(r'${\rm Amplitude\,\,[ppm]}$',fontsize=14)
  ax.set_title(r'${\rm Example}$',fontsize=14)
  plt.show()
