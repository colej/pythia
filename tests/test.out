1.335142414404799638e-06 1.559055997467897647e+01 1.241548142245983959e-03 -9.416511899838546640e-01
0.000000000000000000e+00 2.016413368305810394e+01 1.067954595630112588e-03 -3.118486224688233666e+00
0.000000000000000000e+00 1.678286272220116615e+01 7.550149509782899627e-04 3.104601659682960424e+00
0.000000000000000000e+00 1.428970246127967059e+01 6.363830717854310291e-04 -1.536594474795714316e+00
0.000000000000000000e+00 1.285194531656347294e+01 3.982908042258944672e-04 2.146393754626876316e+00
